src/libstd/sys/unix/rand.rs

```diff
@@ -32,45 +32,14 @@ mod imp {
     use libc;
     use sys::os::errno;
 
-    #[cfg(all(target_os = "linux",
-              any(target_arch = "x86_64",
-                  target_arch = "x86",
-                  target_arch = "arm",
-                  target_arch = "aarch64",
-                  target_arch = "powerpc",
-                  target_arch = "powerpc64",
-                  target_arch = "s390x")))]
+    #[cfg(any(target_os = "linux", target_os = "android"))]
     fn getrandom(buf: &mut [u8]) -> libc::c_long {
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
-        const NR_GETRANDOM: libc::c_long = 0x40000000 + 318;
-        #[cfg(all(target_arch = "x86_64", target_pointer_width = "64"))]
-        const NR_GETRANDOM: libc::c_long = 318;
-        #[cfg(target_arch = "x86")]
-        const NR_GETRANDOM: libc::c_long = 355;
-        #[cfg(target_arch = "arm")]
-        const NR_GETRANDOM: libc::c_long = 384;
-        #[cfg(target_arch = "s390x")]
-        const NR_GETRANDOM: libc::c_long = 349;
-        #[cfg(any(target_arch = "powerpc", target_arch = "powerpc64"))]
-        const NR_GETRANDOM: libc::c_long = 359;
-        #[cfg(target_arch = "aarch64")]
-        const NR_GETRANDOM: libc::c_long = 278;
-
-        const GRND_NONBLOCK: libc::c_uint = 0x0001;
-
         unsafe {
-            libc::syscall(NR_GETRANDOM, buf.as_mut_ptr(), buf.len(), GRND_NONBLOCK)
+            libc::syscall(libc::SYS_getrandom, buf.as_mut_ptr(), buf.len(), libc::GRND_NONBLOCK)
```

Comment:
Why `libc::GRND_NONBLOCK`? Presumably users want this call to be blocking?

Reply:
GRND_NONBLOCK causes getrandom to return an error immediately rather than blocking indefinitely if the kernel randomness source isn't yet initialized. It only matters for things running around init time.

Reply:
The right behavior is to block indefinitely until the kernel randomness source is initialized. See, e.g., [this paper](https://www.usenix.org/system/files/conference/hotos15/hotos15-paper-corrigan-gibbs.pdf). The vast majority of applications that need cryptographic randomness need it to _actually_ be secure. The tiny minority of applications that don't can call `getrandom` directly if they really want to. This is a secure-by-default vs insecure-by-default thing.

Reply:
This API is not publicly exposed. It's only used by `RandomState`.

Reply:
FWIW, Python 3.5 had some controversy about this: https://lwn.net/Articles/693189/

But Python 3.6 does default to blocking now: https://docs.python.org/3.6/library/os.html#os.urandom

Reply:
> This API is not publicly exposed. It's only used by `RandomState`.

Sure, but that implies that `RandomState` would then be insecure. To be concrete on the problem here: if `getrandom` with `GRND_NONBLOCK` returns immediately because the entropy system isn't yet initialized, and then you fall back on reading from `/dev/urandom`, it means that (barring the system getting initialized between the call to `getrandom` and the read from `/dev/urandom`), you'll be reading insecure entropy. 

In fact, it won't just be that single read that will return insecure entropy, it could be all future reads. As section 3.1 of [this paper](https://www.usenix.org/system/files/conference/hotos15/hotos15-paper-corrigan-gibbs.pdf) (see the "Corollary" section under "Reality 1") describes, reading while the entropy system isn't yet initialized can actually allow an attacker to execute attacks that compromise the entropy system on an ongoing basis into the future.

Reply:
There is quite a bit of discussion about this here: https://github.com/rust-lang/rust/issues/32953.

Reply:
OK, fair enough. The TLDR of that discussion seems to be that using `RandomState` for `HashMap`s right after boot is problematic on embedded devices because of the blocking problem.

In keeping with the general trend of default-secure, perhaps we could have blocking be the default, but add a constructor that produces a `RandomState` that doesn't block? Users that need it explicitly could call that constructor, but most users would get the secure option without having to do anything.

Reply:
That doesn't solve the underlying problem, which is that you don't have control of the construction of every HashMap in your dependency graph.

Reply:
Oh, good point. I hadn't thought of that.